receiver/otlpreceiver/metrics/otlp_test.go

```diff
@@ -0,0 +1,279 @@
+					Unit:        "ms",
+					Type:        ocmetrics.MetricDescriptor_CUMULATIVE_INT64,
+					LabelKeys: []*ocmetrics.LabelKey{
+						{Key: "key1"},
+						{Key: "key2"},
+					},
+				},
+				Timeseries: []*ocmetrics.TimeSeries{
+					{
+						StartTimestamp: unixnanoToTimestamp(unixnanos1),
+						LabelValues: []*ocmetrics.LabelValue{
+							{
+								// key1
+								Value:    "value1",
+								HasValue: true,
+							},
+							{
+								// key2
+								HasValue: false,
+							},
+						},
+						Points: []*ocmetrics.Point{
+							{
+								Timestamp: unixnanoToTimestamp(unixnanos2),
+								Value: &ocmetrics.Point_Int64Value{
+									Int64Value: 123,
+								},
+							},
+						},
+					},
+					{
+						StartTimestamp: unixnanoToTimestamp(unixnanos1),
+						LabelValues: []*ocmetrics.LabelValue{
+							{
+								// key1
+								HasValue: false,
+							},
+							{
+								// key2
+								Value:    "value2",
+								HasValue: true,
+							},
+						},
+						Points: []*ocmetrics.Point{
+							{
+								Timestamp: unixnanoToTimestamp(unixnanos2),
+								Value: &ocmetrics.Point_Int64Value{
+									Int64Value: 456,
+								},
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+
+	assert.EqualValues(t, metricsData, metricSink.metricsDataList[0])
+}
+
+func makeMetricsServiceClient(port int) (collectormetrics.MetricsServiceClient, func(), error) {
+	addr := fmt.Sprintf(":%d", port)
+	cc, err := grpc.Dial(addr, grpc.WithInsecure(), grpc.WithBlock())
+	if err != nil {
+		return nil, nil, err
+	}
+
+	metricsClient := collectormetrics.NewMetricsServiceClient(cc)
+
+	doneFn := func() { _ = cc.Close() }
+	return metricsClient, doneFn, nil
+}
+
+func otlpReceiverOnGRPCServer(t *testing.T, mc consumer.MetricsConsumer) (r *Receiver, port int, done func()) {
+	ln, err := net.Listen("tcp", "localhost:")
+	require.NoError(t, err, "Failed to find an available address to run the gRPC server: %v", err)
+
+	doneFnList := []func(){func() { ln.Close() }}
+	done = func() {
+		for _, doneFn := range doneFnList {
+			doneFn()
+		}
+	}
+
+	_, port, err = hostPortFromAddr(ln.Addr())
+	if err != nil {
+		done()
+		t.Fatalf("Failed to parse host:port from listener address: %s error: %v", ln.Addr(), err)
+	}
+
+	r, err = New(receiverTagValue, mc)
+	require.NoError(t, err, "Failed to create the Receiver: %v", err)
+
+	// Now run it as a gRPC server
+	srv := observability.GRPCServerWithObservabilityEnabled()
+	collectormetrics.RegisterMetricsServiceServer(srv, r)
+	go func() {
+		_ = srv.Serve(ln)
+	}()
+
+	return r, port, done
+}
+
+func hostPortFromAddr(addr net.Addr) (host string, port int, err error) {
+	addrStr := addr.String()
+	sepIndex := strings.LastIndex(addrStr, ":")
+	if sepIndex < 0 {
+		return "", -1, errors.New("failed to parse host:port")
+	}
+	host, portStr := addrStr[:sepIndex], addrStr[sepIndex+1:]
+	port, err = strconv.Atoi(portStr)
+	return host, port, err
+}
+
+type metricAppender struct {
```

Comment:
This probably belongs to a receivertest helper package under receivers (we already have a similar one for exporters in exporters/exportertest and for processors).

Reply:
Extracted.